operator/k8s_cep_gc.go

```diff
@@ -64,14 +80,37 @@ func enableCiliumEndpointSyncGC() {
 							Errorf("Saw %T object while expecting *cilium_v2.CiliumEndpoint", cepObj)
 						continue
 					}
-
 					cepFullName := cep.Namespace + "/" + cep.Name
-					_, exists, err := watchers.PodStore.GetByKey(cepFullName)
-					if err != nil {
-						scopedLog.WithError(err).Warn("Unable to get pod from store")
-						continue
+
+					// If we are running this function "once" it means that we
+					// will delete all CEPs in the cluster regardless of the pod
+					// state therefore we won't even watch for the pod store.
+					delCEP := once
+					if !once {
+						podObj, exists, err := watchers.PodStore.GetByKey(cepFullName)
+						if err != nil {
+							scopedLog.WithError(err).Warn("Unable to get pod from store")
+							continue
+						}
+						if !exists {
+							delCEP = true
+						} else {
+							pod := podObj.(*slim_corev1.Pod)
+							if !ok {
+								log.WithField(logfields.Object, podObj).
+									Errorf("Saw %T object while expecting *slim_corev1.Pod", podObj)
+								continue
+							}
+							// In Kubernetes Jobs, Pods can be left in Kubernetes until the Job
+							// is deleted. If the Job is never deleted, Cilium will never receive a Pod
+							// delete event, causing the IP to be left in the ipcache.
+							// For this reason we should delete the ipcache entries whenever the pod
+							// status is either PodFailed or PodSucceeded as it means the IP address
+							// is no longer in use.
+							delCEP = !k8sUtils.IsPodRunning(pod.Status)
+						}
```

Comment:
Similar remark to Chris, I find the code is becoming a bit more complicated with the different branches. Maybe it would be worth moving the computation of `delCEP` to its own function? That's orthogonal to the addition of branches but may help reading through.